### PR TITLE
Improve stats performance

### PR DESCRIPTION
**Pull request by planarvoid**
There has been some obvious UI thread blocking in the stats that caused the whole UI to freeze while loading the stats for the first time. I decided to look into it and the main reason I've found is that the `buildUiModel` operation happens on the main thread. This is caused by using LiveData (and merge on live data) the same way we'd use RxJava streams. However, this is wrong. 

The change I've implemented instead is taking the 3 instances of live data (model, domain state and ui state), which were merged and mapped to UI model, and replacing them by a coroutine approach. Now instead of directly calling `uiState.postValue(newState)` we'd rather just set the field and call an `updateState` method which is asynchronous. This method builds the ui model and when it's done, it posts it to the exposed LiveData instance. 

To test:
* Stress test Stats - leaving, refreshing, failures

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by loremattei** [comment]
I'm moving this to 13.3 since 13.2 has been cut. If you want this to make it to 13.2, please feel free to move it back, target the release branch and ping me to build a new beta.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/detail/StatsDetailViewModel.kt [hunk 1]
```diff
@@ -50,9 +51,9 @@ class StatsDetailViewModel
         statsPostProvider.init(postId, postType, postTitle, postUrl)
     }
 
-    fun onDateChanged() {
+    fun onDateChanged(selectedSection: StatsSection) {
```
Comment (0nko) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/detail/StatsDetailViewModel.kt]: 💯

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt [hunk 2]
```diff
@@ -200,6 +191,43 @@ abstract class BaseStatsUseCase<DOMAIN_MODEL, UI_STATE>(
         return buildLoadingItem() + listOf(BlockListItem.Empty(textResource = R.string.stats_no_data_yet))
     }
 
+    private fun updateState() {
+        updateJob?.let { job ->
+            if (job.isActive) {
+                job.cancel()
+            }
+        }
+        updateJob = GlobalScope.launch(backgroundDispatcher) {
```
Comment (0nko) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt]: Is the `GlobalScope` necessary here? Wouldn't it be possible to pass down a scope from the ViewModel?

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/BaseListUseCase.kt [hunk 3]
```diff
@@ -40,15 +42,15 @@ class BaseListUseCase(
...
+    val data: MediatorLiveData<UiModel> = mergeAsyncNotNull(bgDispatcher, statsTypes, blockListData) { types, map ->
         types.mapNotNull {
             if (map.containsKey(it)) {
                 map[it]
             } else {
                 null
             }
         }
-    }.map { useCaseModels ->
+    }.mapAsync(bgDispatcher) { useCaseModels ->
```
Comment (0nko) [inline WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/BaseListUseCase.kt]: 👍 I wonder why we didn't think of making this async sooner 😃.

**Review by 0nko — APPROVED** [review]
This looks excellent, Vojta! 🏅

The changes make sense and the performance looks great. So I really have no comments, just one question in the code.

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/BaseListUseCase.kt (hunk 3 above)
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/BaseListUseCase.kt]: for me it was the "this is like RxJava" approach where I considered everything in the stream to run on a background thread :-) - an expensive mistake

File: WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt (hunk 2 above)
Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt]: I'll look into it, it's a great suggestion 👍

Comment (planarvoid) [reply WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt]: I've changed the `BaseStatsUseCase` and the `BaseListUseCase` to implement CoroutineScope and use that instead. Their life cycle is longer than `ViewModel` so I'm not cancelling the jobs now. I can look into that in another PR.

**Comment by planarvoid** [comment]
@0nko thanks for the review!  I've spent some time with more optimisation and fixed one bug that wasn't apparent before (wrong empty state in the DWMY tabs). Now the stats were sometimes showing the Insights empty screen when loading data. 
It should be ready for another pass

**Review by 0nko — APPROVED** [review]
Thanks for the changes, looks great!

**Pull request merged**